File: backend/app/models/import_job.py
"""
数据导入任务模型

记录 Excel/CSV 数据导入任务的状态、结果和错误信息
"""

from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Index, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from enum import Enum

from app.models.base import Base, IDMixin, TimestampMixin


class ImportSourceType(str, Enum):
    """导入源类型"""
    EXCEL = "excel"
    CSV = "csv"


class ImportTargetType(str, Enum):
    """导入目标类型"""
    ORDERS = "orders"
    EXPENSE_RECORDS = "expense_records"
    STORES = "stores"
    EXPENSE_TYPES = "expense_types"


class ImportJobStatus(str, Enum):
    """导入任务状态"""
    PENDING = "pending"           # 待处理
    RUNNING = "running"           # 运行中
    SUCCESS = "success"           # 全部成功
    PARTIAL_FAIL = "partial_fail" # 部分失败
    FAIL = "fail"                 # 全部失败


class DataImportJob(Base, IDMixin, TimestampMixin):
    """数据导入任务表"""
    
    __tablename__ = "data_import_jobs"
    
    job_name = Column(String(200), nullable=False, comment="任务名称")
    source_type = Column(
        SQLEnum(ImportSourceType, name="import_source_type", create_type=False),
        nullable=False,
        comment="源文件类型 (excel/csv)"
    )
    target_type = Column(
        SQLEnum(ImportTargetType, name="import_target_type", create_type=False),
        nullable=False,
        comment="目标数据类型 (orders/expense_records/stores/expense_types)"
    )
    status = Column(
        SQLEnum(ImportJobStatus, name="import_job_status", create_type=False),
        nullable=False,
        default=ImportJobStatus.PENDING,
        comment="任务状态"
    )
    
    # 文件信息
    file_name = Column(String(500), nullable=False, comment="原始文件名")
    file_path = Column(String(1000), nullable=False, comment="文件存储路径")
    
    # 导入结果统计
    total_rows = Column(Integer, nullable=False, default=0, comment="总行数")
    success_rows = Column(Integer, nullable=False, default=0, comment="成功行数")
    fail_rows = Column(Integer, nullable=False, default=0, comment="失败行数")
    
    # 错误报告
    error_report_path = Column(String(1000), nullable=True, comment="错误报告文件路径")
    
    # 额外配置（存储字段映射、门店ID等）
    config = Column(JSONB, nullable=True, comment="任务配置 (映射关系、门店ID等)")
    
    # 创建用户
    created_by_id = Column(Integer, ForeignKey("user.id", ondelete="SET NULL"), nullable=True, comment="创建用户ID")
    
    # 关系
    created_by = relationship("User", foreign_keys=[created_by_id])
    errors = relationship("DataImportJobError", back_populates="job", cascade="all, delete-orphan")
    
    # 索引
    __table_args__ = (
        Index("idx_import_job_status", "status"),
        Index("idx_import_job_target_type", "target_type"),
        Index("idx_import_job_created_at", "created_at"),
        Index("idx_import_job_created_by", "created_by_id"),
    )
    
    def __repr__(self):
        return f"<DataImportJob(id={self.id}, job_name='{self.job_name}', status='{self.status}')>"


class DataImportJobError(Base, IDMixin, TimestampMixin):
    """数据导入错误记录表"""
    
    __tablename__ = "data_import_job_errors"
    
    job_id = Column(Integer, ForeignKey("data_import_jobs.id", ondelete="CASCADE"), nullable=False, comment="任务ID")
    row_no = Column(Integer, nullable=False, comment="错误行号 (从1开始)")
    field = Column(String(100), nullable=True, comment="错误字段")
    message = Column(Text, nullable=False, comment="错误信息")
    raw_data = Column(JSONB, nullable=True, comment="原始行数据")
    
    # 关系
    job = relationship("DataImportJob", back_populates="errors")
    
    # 索引
    __table_args__ = (
        Index("idx_import_error_job_id", "job_id"),
        Index("idx_import_error_row_no", "job_id", "row_no"),
    )
    
    def __repr__(self):
        return f"<DataImportJobError(id={self.id}, job_id={self.job_id}, row_no={self.row_no})>"

File: frontend/src/types/modules/import_job.ts
/**
 * 数据导入相关类型定义
 */

// 导入来源类型（对应后端 ImportSourceType）
export enum ImportSourceType {
  EXCEL = 'excel',
  CSV = 'csv'
}

// 导入目标类型（对应后端 ImportTargetType）
export enum ImportTargetType {
  ORDERS = 'orders',
  EXPENSE_RECORDS = 'expense_records',
  STORES = 'stores',
  EXPENSE_TYPES = 'expense_types'
}

// 导入任务状态（对应后端 ImportJobStatus）
export enum ImportJobStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  SUCCESS = 'success',
  PARTIAL_FAIL = 'partial_fail',
  FAIL = 'fail'
}

// 导入任务
export interface ImportJob {
  id: number
  filename: string
  source_type: ImportSourceType
  target_type: ImportTargetType
  status: ImportJobStatus
  store_id?: number
  store_name?: string
  total_rows: number
  success_rows: number
  fail_rows: number
  created_by_id: number
  created_by_name?: string
  created_at: string
  started_at?: string
  finished_at?: string
}

// 导入任务详情（包含用户信息）
export interface ImportJobDetail extends ImportJob {
  file_path: string
  error_report_path?: string
  config?: Record<string, unknown>
  created_by?: {
    id: number
    username: string
    real_name?: string
  }
}

// 导入错误记录
export interface ImportJobError {
  id: number
  job_id: number
  row_number: number
  error_type: string
  error_message: string
  raw_data?: Record<string, unknown>
  created_at: string
}

// 创建导入任务请求
export interface ImportJobCreateRequest {
  file: File
  target_type: ImportTargetType
  store_id?: number
}

// 导入任务查询参数
export interface ImportJobQuery {
  page: number
  page_size: number
  target_type?: ImportTargetType
  status?: ImportJobStatus
  start_date?: string
  end_date?: string
  keyword?: string
}

// 导入错误查询参数
export interface ImportJobErrorQuery {
  page: number
  page_size: number
}

// 状态显示映射
export const ImportJobStatusMap = {
  [ImportJobStatus.PENDING]: { text: '待处理', color: 'info', type: 'info' as const },
  [ImportJobStatus.RUNNING]: { text: '运行中', color: 'warning', type: 'warning' as const },
  [ImportJobStatus.SUCCESS]: { text: '全部成功', color: 'success', type: 'success' as const },
  [ImportJobStatus.PARTIAL_FAIL]: { text: '部分失败', color: 'warning', type: 'warning' as const },
  [ImportJobStatus.FAIL]: { text: '全部失败', color: 'danger', type: 'danger' as const }
}

// 目标类型显示映射
export const ImportTargetTypeMap = {
  [ImportTargetType.ORDERS]: '订单数据',
  [ImportTargetType.EXPENSE_RECORDS]: '费用记录',
  [ImportTargetType.STORES]: '门店信息',
  [ImportTargetType.EXPENSE_TYPES]: '费用科目'
}

// 创建任务时可选的导入类型（仅包含后端已实现）
export const ImportTargetTypeCreateMap = {
  [ImportTargetType.ORDERS]: '订单数据',
  [ImportTargetType.EXPENSE_RECORDS]: '费用记录'
}

// 来源类型显示映射
export const ImportSourceTypeMap = {
  [ImportSourceType.EXCEL]: 'Excel',
  [ImportSourceType.CSV]: 'CSV'
}
